mobile/mobile.go

```diff
@@ -0,0 +1,492 @@
+
+	// set debug level to 'info' if not specified
+	if debugLevel == "" {
+		debugLevel = "info"
+	}
+
+	// Hook interceptor for os signals.
+	shutdownInterceptor, err := signal.Intercept()
+	if err != nil {
+		exit(err)
+	}
+
+	logWriter := build.NewRotatingLogWriter()
+	SetupLoggers(logWriter, shutdownInterceptor)
+
+	err = build.ParseAndSetDebugLevels(debugLevel, logWriter)
+	if err != nil {
+		exit(err)
+	}
+
+	m, err = newMobileClient()
+	if err != nil {
+		exit(fmt.Errorf("config validation error: %v", err))
+	}
+
+	for _, registration := range registrations {
+		registration(registry)
+	}
+
+	log.Debugf("Mobile client ready for connecting")
+}
+
+func newMobileClient() (*mobileClient, error) {
+	return &mobileClient{
+		lndConn: nil,
+		registry: make(map[string]func(context.Context,
+			*grpc.ClientConn, string, func(string, error))),
+	}, nil
+}
+
+func ConnectServer(mailboxServer string, isDevServer bool,
+	pairingPhrase string, localStatic string, remoteStatic string) {
+	if mailboxServer == "" {
+		exit(fmt.Errorf("invalid use of ConnectServer, " +
+			"need parameter mailboxServer"))
+	}
+
+	// Check that the correct arguments and config combinations have been
+	// provided.
+
+	err := validateArgs(localStatic, remoteStatic)
+	if err != nil {
+		exit(err)
+	}
+
+	// Parse the key arguments.
+	localPriv, remotePub, err := parseKeys(
+		localStatic, remoteStatic,
+	)
+	if err != nil {
+		exit(err)
+	}
+
+	// Disable TLS verification for the REST connections if this is a dev
+	// server.
+	if isDevServer {
+		defaultHttpTransport := http.DefaultTransport.(*http.Transport)
+		defaultHttpTransport.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: true,
+		}
+	}
+
+	statusChecker, lndConnect, err := core.MailboxRPCConnection(
+		mailboxServer, pairingPhrase, localPriv, remotePub,
+		func(key *btcec.PublicKey) error {
+			remoteKeyReceiveCallback.SendResult(hex.EncodeToString(
+				key.SerializeCompressed(),
+			))
+
+			return nil
+		}, func(data []byte) error {
+			parts := strings.Split(string(data), ": ")
+			if len(parts) != 2 || parts[0] != "Macaroon" {
+				return fmt.Errorf("authdata does " +
+					"not contain a macaroon")
+			}
+
+			macBytes, err := hex.DecodeString(parts[1])
+			if err != nil {
+				return err
+			}
+
+			mac := &macaroon.Macaroon{}
+			err = mac.UnmarshalBinary(macBytes)
+			if err != nil {
+				return fmt.Errorf("unable to decode "+
+					"macaroon: %v", err)
+			}
+
+			m.mac = mac
+
+			authDataCallback.SendResult(string(data))
+
+			return nil
+		},
+	)
+
+	if err != nil {
+		exit(err)
+	}
+
+	m.statusChecker = statusChecker
+	m.lndConn, err = lndConnect()
+	if err != nil {
+		exit(err)
+	}
+
+	log.Debugf("Mobile client connected to RPC")
+}
+
+func IsConnected() bool {
+	return m.lndConn != nil
+}
+
+func Disconnect() {
+	if m.lndConn != nil {
+		if err := m.lndConn.Close(); err != nil {
+			log.Errorf("Error closing RPC connection: %v", err)
+		}
+		m.lndConn = nil
+	}
+}
+
+func Status() string {
```

Comment:
Related to my comment above: Couldn't this be a method on a client instance instead? Or do we need to enforce a singleton pattern for some reason?

Reply:
Ah, I guess this was just taken from the WASM client where we only ever have a single instance per process... So not a huge priority to refactor this.

Reply:
ah, seeing this comment now. I agree the methods should be on a client instance instead. The wasm client does allow for multiple connections per process as it stands (differentiated by namespaces) So I think we should do the same here? Generally nice to avoid global variables